examples/api/lib/material/menu_bar/create_cascading_menu.0.dart

```diff
@@ -0,0 +1,230 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// Flutter code sample for [createCascadingMenu].
+
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+const String kMessage = '"Talk less. Smile more." - A. Burr';
+
+void main() => runApp(const MenuBarApp());
+
+enum MenuSelection {
```

Comment:
I was curious what this all would look like if MenuSelection was an ordinary class that bundled the menu item's callback or submenu.  I just typed in all of the code that follows. Did not compile etc. Hopefully it's trivial enough to be self-explanatory.

A simple definition of MenuSelection as an ordinary class:

```dart
class MenuSelection {
  const class MenuSelection({ required this.label, this.shortcut, this.onSelected, this.menu });

  final String label;
  final MenuSerializableShortcut? shortcut;
  final VoidCallback onSelected
  final List<MenuSelection>? menu;

  static List<Widget> build(List<MenuSelection> selections) {
    Widget buildSelection(MenuSelection selection) {
      if (selection.menu != null) {
        return MenuButton(
          label: Text(selection.label),
          children: MenuSelection.build(selection.menu),
        );
      }
      return MenuItemButton(
        label: Text(selection.label),
        shortcut: selection.shortcut,
        onSelected: selection.onSelected,
      );
    }
    return selections.map<Widget>(buildSelection).toList();
  }

  static List<MenuSerializableShortcut> shortcuts(List<MenuSelection> selections) {
    List<MenuSerializableShortcut> value = MenuSerializableShortcut[];
    for (final MenuSelection selection in selections) {
      if (selection.menu != null) {
        value.addAll(MenuSelection.shorts(selection.menu));
      } else {
        if (selection.shortcut != null) {
          value.add(selection.shortcut);
        }
      }
    }
  }
}
``` 

The static methods would be used in init() and didUpdateWidget(). For example:
```dart
shortcutsEntry = ShortcutRegistry.of(context).addAll(MenuSelection.shortcuts(selections));
```
The calls to updateMenuEntry would be replaced by:
```dart
menuEntry.children = MenuSelection.build(selections);
```

And `selections` would be defined in the State subclass:
```dart
final List<MenuSelection> selections = <MenuSelection>[
  MenuSelection(
    label: 'About',
    onSelected: () {
      showAboutDialog(
        context: context,
        applicationName: 'MenuBar Sample',
        applicationVersion: '1.0.0',
      ),
    },
  ),
  MenuSelection(
    label: 'Show Message',
    shortcut: SingleActivator(LogicalKeyboardKey.keyS, control: true)
    onSelected: () {
      setState(() { showingMessage = !showingMessage; });
    },
  ),
  MenuSelection(
    label: 'Reset Message',
    shortcut: SingleActivator(LogicalKeyboardKey.escape),
    onSelected: !showingMessage ? null : () {
      setState() { showingMessage = false; };
    },
  ),
  MenuSelection(
    label: 'Hide Message'
    onSelected: () {
      setState(() { showingMessage = false; }),
    },
  ),
  MenuSelection(
    label: 'Background Color',
    menu: <MenuSelection>[
      MenuSelection(
        label: 'Red Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyR, control: true),
        onSelected: () { backgroundColor = Colors.red; },
      ),
      MenuSelection(
        label: 'Green Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyG, control: true),
        onSelected: () { backgroundColor = Colors.green; },
      ),
      MenuSelection(
        label: 'Blue Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyB, control: true),
        onSelected: () { backgroundColor = Colors.blue; },
      ),
    ),
  ),
];
```
This has the advantage of keeping the menu structure and the menus actions in one definition. The disadvantage is that MenuSelection is still a utility abstraction and in this version it hides the mapping from menu items to MenuItem and MenuItemButton.

Reply:
If I understand this correctly, this is moving back in the direction of menus that are defined by non-widgets, and I thought we moved away from that precisely because they weren't widgets. I mean, now you're not prevented from making non-MenuItemButtons/MenuButtons, but you'd have to write your own equivalent to the `build` static method here to do it.

I'm getting the feeling that I must be missing something here, though.

Also, what's to stop someone from writing something like this on top of the code in this PR?

Reply:
Currently the example menu's implementation is spread across the app-specific MenuSelection enum class, the _activate() method, and the _updateMenuEntry method.  My goal was to reduce that by one: to a reusable MenuSelection class and the initializer for the selections state variable. The MenuSelection class could be copy-and-pasted and shared by all menus in an application. And yes, all of this can be built on top of the code in the PR, I didn't mean to suggest that it should be part of the new API.

Some mechanism that bridges between the shortcuts and the subset of  menu item onSelected callbacks that support those shortcuts is needed.  There are definitely other possibilities, for example the shared callbacks could be defined as methods within the State subclass and shared with the corresponding shortcut definitions.  I think it's fine to let idiomatic approaches to defining menus and shortcuts emerge as they see use in real apps.

Reply:
Ahh, right, this is for the example, sorry, I missed that.  Sure, I'll see if I can consolidate in the way you suggest.

Reply:
Note also: this example is a special (common?) case where the menus are defined exclusively by MenuButtons and MenuItemButtons. It would probably be possible to write a utility function that registered the shortcuts by just walking the widget tree each time it was rebuilt.

Reply:
OK, I modified the `menu_bar.0.dart` example to follow this model. I'm not sure it's "ideal" yet, but it gives another example of how to define menus. I left the other examples using enhanced enums so that they can see multiple forms.